cpp/src/quantiles/tdigest/tdigest.cu

```diff
@@ -0,0 +1,279 @@
+  children.push_back(make_empty_column(data_type(type_id::FLOAT64)));
+
+  auto offsets = cudf::make_fixed_width_column(
+    data_type(type_id::INT32), 2, mask_state::UNALLOCATED, stream, mr);
+  thrust::fill(rmm::exec_policy(stream),
+               offsets->mutable_view().begin<offset_type>(),
+               offsets->mutable_view().begin<offset_type>() + 2,
+               0);
+  return make_lists_column(1,
+                           std::move(offsets),
+                           cudf::make_structs_column(0, std::move(children), 0, {}, stream, mr),
+                           0,
+                           {});
+}
+
+std::unique_ptr<column> percentile_approx(column_view const& input,
+                                          column_view const& percentages,
+                                          rmm::cuda_stream_view stream,
+                                          rmm::mr::device_memory_resource* mr)
+{
+  check_is_valid_tdigest_column(input);
+
+  // output is a list column with each row containing percentages.size() percentile values
+  auto offsets = cudf::make_fixed_width_column(
+    data_type{type_id::INT32}, input.size() + 1, mask_state::UNALLOCATED, stream, mr);
+  auto row_size_iter = thrust::make_constant_iterator(percentages.size());
+  thrust::exclusive_scan(rmm::exec_policy(stream),
+                         row_size_iter,
+                         row_size_iter + input.size() + 1,
+                         offsets->mutable_view().begin<offset_type>());
+
+  // if any of the input digests are empty, nullify the corresponding output rows (values will be
+  // uninitialized)
+  auto [bitmask, null_count] = [stream, mr, input]() {
+    lists_column_view lcv(input);
+    auto iter = cudf::detail::make_counting_transform_iterator(
+      0, [offsets = lcv.offsets().begin<offset_type>()] __device__(size_type index) {
+        return offsets[index + 1] - offsets[index] == 0 ? 1 : 0;
+      });
+    auto const null_count = thrust::reduce(rmm::exec_policy(stream), iter, iter + input.size(), 0);
+    if (null_count == 0) {
+      return std::pair<rmm::device_buffer, size_type>{rmm::device_buffer{}, null_count};
+    }
+    return cudf::detail::valid_if(
+      thrust::make_counting_iterator(0),
+      thrust::make_counting_iterator(0) + input.size(),
+      [offsets = lcv.offsets().begin<offset_type>()] __device__(size_type index) {
+        return offsets[index + 1] - offsets[index] == 0 ? 0 : 1;
+      },
+      stream,
+      mr);
+  }();
+
+  return cudf::make_lists_column(input.size(),
+                                 std::move(offsets),
+                                 compute_approx_percentiles(input, percentages, stream, mr),
+                                 null_count,
+                                 std::move(bitmask),
+                                 stream,
+                                 mr);
+}
+
+}  // namespace detail
+
+std::unique_ptr<column> percentile_approx(
```

Comment:
The default value for `mr` here reminds me: Should this not go into a header, for use from JNI/`spark-rapids`?